### PR TITLE
Tesla Model S support

### PR DESCRIPTION
**Pull request by geohot**
Model S Support by:
Jean-Claude Thibault (Kalud)
and
Robert Cotran (zax123)

A huge thanks to them! The openpilot community is amazing.

Check out a video of it working on a pre AP Tesla:
https://www.youtube.com/watch?v=ro48vcN4VxY

Note: All new car support is extremely alpha. Unless you are ready to spend time tinkering and know how to safely experiment, please wait until this is merged into openpilot master and been audited for safety.

### CODE REVIEW

**Comment by jeankalud** [comment]
Hardware instructions: [openpilot.robertcotran.com](http://openpilot.robertcotran.com/)

**Comment by jeankalud** [comment]
Two full 6 min without touching the steering wheel (except to bump the 6 min user awareness): https://youtu.be/Q2diuQ53Too

Also, quick timelapse: https://youtu.be/ZBTv1Yfjgvg

And longer video (30 min) super boring!: https://youtu.be/H4nVV9NhLl0

**Comment by neilsambhu** [comment]
Hello everyone,
I am new to github and the developments surrounding comma.ai and openpilot; I would love to learn more about self-driving technology. I have some questions regarding the hardware instructions (openpilot.robertcotran.com) posted by @jeankalud:

1. Can the Neo be interchanged for the EON Dashcam DevKit (https://shop.comma.ai/products/eon-dashcam-devkit)? If not, would the neodriven be suitable (https://www.neodriven.com/)?
2. Is the panda OBD-II Dongle (https://shop.comma.ai/products/panda-obd-ii-dongle) a suitable interface to the diagnostic connector to tap into the chassis CAN, as mentioned in the "Establishing a connection to the car's chassis CAN bus" section of the hardware instructions?

If there is a better way for me to contact Jean-Claude Thibault and Robert Cotran, please let me know. 

Thanks

**Comment by jeankalud** [comment]
Join #tesla on comma.slack.com

Short answer, yes to both questions.

File: selfdrive/car/tesla/teslacan.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+import struct
+
+import common.numpy_fast as np
+from selfdrive.config import Conversions as CV
+
+
+# *** Honda specific ***
```
Comment (klaus385) [inline selfdrive/car/tesla/teslacan.py]: Is this block still needed for Tesla port? Since its supposed to be "Honda Specific"?

Comment (jeankalud) [reply selfdrive/car/tesla/teslacan.py]: This is unused. There will be a much cleaner pull request with updated to 0.3.8.2 codebase in about two weeks. I am currently porting the code and testing.

**Comment by max0vdh** [comment]
jeankalud, thank you for your efforts.  I am assembling hardware to test this out.  It appears that the EPAS bypass cable you describe omits the BATT and GND connections to the EPAS, just leaving them unpopulated on the X164 side.  Am I misreading the instructions or is that really intended? Also, I plan to include an in-cabin DPDT switch to un-bypass the kill signal so I can easily remove the NEO temporarily.  I want to make sure the EPAS is reconnected as originally designed in that mode.   

I wish you luck learning how to communicate with a later model radar unit as well to incorporate TACC.  Wish I could be of help with that.

**Comment by jeankalud** [comment]
Batt (ignition) and ground are passed from the male to female connectors at the EPAS. The concept is to plug the unused connector from the car’s harness to protect from elements, but we still use its ignition and ground signals. That way the cable going into the cabin only has to carry the canbus signals going to the EPAS. 

We do use a DPDT switch in the cabin becaus you need to restore original signals from car’s chassis canbus to drive without the NEO/EON, otherwise you get “Reduced steering power” safe mode on the EPAS.

Contact us in the slack channel is needed, we’ll get you running ;) It a bit intimidating at first but overall not that difficult and totally reversible. And it works so well!

**Comment by max0vdh** [comment]
@jeankalud thank you for your response.  Everything about your bypass cable strategy seems sensible but specifically, I was questioning whether pin 1 (batt) and pin 2 (gnd) should be passed directly to the EPAS as well.  It is clear that ignition (pin 6) and CAN- (pin 2) get there.  Sorry if I'm missing something obvious.

**Comment by bitshop** [comment]
@geohot What do we need to do to get this merged, other than obvious like get it to latest OP version. The cars don't appear to have computer controllable brakes, so that's a limitation. The fingerprint and other things are missing here, so there's some obvious work.

I think it should run on Panda/Eon also as a qualification for 0.4.x. 

I suspect we MUST get longitudinal control integrated with the cruise system at some point, but as this drifts further from merge-able that creates a barrier for the people working on this to fork instead of merge, that doesn't seem like a desired outcome. 

What's the bar you set for getting this merged? Perhaps if high we can merge a working version but with a fingerprint commented out explaining why you must understanding how truly alpha it is - But that way we eliminate the code base drifting from the pull request.

**Comment by sveip** [comment]
@jeankalud, your page states _The classic Model S does not have autopilot hardware but does have the same steering control system and servo as the autopilot Model S and therefore can be controlled through electrical signaling (specifically one of the car's CAN buses)._

Are you sure this is the case? I used to have an 2013 S and now have an 2015 AP S. The 2013 pre-AP had another steering rack and controls, but later models (mid 2014, pre AP) were delivered with newer steering rack but w/o radar and camera.

Effectively, I think this means that OpenPilot will only work on the later pre-AP cars.

Can you confirm/decline this?

**Comment by jeankalud** [comment]
@sveip My car is a November 2013 pre-AP and I've driven more than 20,000 km on Openpilot already. Also, my friend zax123 has a March 2013 Model S and is also using Openpilot every days. Lateral control only for us. The commands to control the steering angle are the same on newer (2018) and older (2012-2013) Model S, but the steering rack hardware did change at some point, like you pointed. Regardless, all those electric power steering are controllable. Please join the Slack channel for more discussions: slack.comma.ai

**Comment by jnrowell** [comment]
I'd like to get openpilot running on my 2013 Model S. However I could not figure out how to join the Slack channel - it's asking me for a teslamotors.com email address, which I don't have. Can you provide a list of the parts that are needed? I've looked at the linked sites but they're not making sense to me for some reason, and they appear to have not been updated since last October. Would appreciate some direction here, thanks!

**Comment by andy867** [comment]
Join at comma.slack.com and login with Google account

**Comment by jnrowell** [comment]
I tried entering my gmail address but what's the password? I tried the "forgot password" link and I got an email saying my email address "doesn't have an account for that workspace". So how can I get an account?? Here's what I see when I try to log in:
screenshot

**Comment by jnrowell** [comment]
When I clicked the "create an account" link, this is what I got:
screenshot2

So I guess I need an invitation. Dear admin, may I have an invite?

**Comment by jnrowell** [comment]
OK, I think I finally figured it out. I'll put this out there for others who are having the same trouble I had. You can't join directly on comma.slack.com - you first have to visit https://commaslack.herokuapp.com/ and enter your email address. Then once you're signed up, go to comma.slack.com and log in.

**Comment by Palvaran** [comment]
@jnrowell, Thanks. I was wondering about this as well.

**Comment by geohot** [comment]
See #246

**Pull request closed without merging**